src/modules/roc_netio/target_uv/roc_netio/transceiver.cpp

```diff
@@ -303,112 +300,118 @@ void Transceiver::process_tasks_() {
 }
 
 bool Transceiver::add_udp_receiver_(Task& task) {
-    if (stopped_) {
-        roc_log(LogError, "transceiver: can't add port %s: transceiver is stopped",
-                packet::address_to_str(*task.address).c_str());
-        return false;
-    }
+    roc_panic_if(stopped_);
 
-    if (has_port_(*task.address)) {
+    if (get_port_(*task.address)) {
```

Comment:
Do we need this check? Can we instead rely on OS to handle duplicate addresses?

* If we don't need it, we can remove it.
* If we need it, then 1) could you explain why? 2) we have a problem with current implementation.

The problem.

Step 1:

* the user calls add_udp_receiver("0.0.0.0:12345")
* trx creates UDPReceiverPort and adds it to ports_ list
* trx calls UDPReceiverPort::open("0.0.0.0:12345") but it fails for some reason and returns before setting UDPReceiverPort::address_
* trx calls UDPReceiverPort::async_close()
* the asynchronous close is still in progress...

Step 2:

* the user calls add_udp_receiver("0.0.0.0:0") (zero port means random port)
* trx calls get_port_("0.0.0.0:0") and it returns the UDPReceiverPort constructed at step 1 (because its address is not initialized and so is also "0.0.0.0:0")
* trx incorrectly complains that the port already exists and returns the user an error

If can't just remove the check, we can:

* ensure that address_ is already initialized (as in your previous revision)
* OR (it's probably a cleaner approach) place ports that are uninitialized/being closed into a separate list

Reply:
We probably have this check because Transceiver::remove_port() assumes that there is always only one port for given address.

This approach will fail when we'll add TCP (because Address is IP + port, but it has no info about protocol - TCP or UDP).

So it seems that we anyway should rework the remove_port() interface somehow. If this is the only reason why we have this check, we can just do it now.

Reply:
> Do we need this check? Can we instead rely on OS to handle duplicate addresses?

Yes, we can. This check was introduced before this PR, thus I was trying to
keep the same behaviour.

> trx calls get_port_("0.0.0.0:0") and it returns the UDPReceiverPort
> constructed at step 1 (because its address is not initialized and so is also "0.0.0.0:0")

It is another problem, isn't it? I would expect that zero value for address
means invalid address, thus port also should have an invalid value.

> This approach will fail when we'll add TCP (because Address is IP + port, but it has no info about protocol - TCP or UDP).

This approach won't fail if we add an info about an address wheather it TCP or UDP,
thus the following check will find a correct address:

```cpp
core::SharedPtr<BasicPort> Transceiver::get_port_(const packet::Address& address) const {
    for (core::SharedPtr<BasicPort> pp = ports_.front(); pp; pp = ports_.nextof(*pp)) {
        if (pp->address() == address) {
            return pp;
        }
    }

    return NULL;
}
```

Now we can do the following:

- rely on OS to handle duplicate address
- leave remove_port() implementation as is and rework address structure later